cluster/ci/origin/logging-infra/gcs.go

```diff
@@ -0,0 +1,82 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"strconv"
+	"sync"
+
+	"cloud.google.com/go/storage"
+)
+
+// Metadata holds run-time data about the build
+type Metadata struct {
+	// JobName is the name of the job triggering
+	// this build
+	JobName string `json:"job-name"`
+
+	// BuildNumber is the identifier for this build
+	// of the job
+	BuildNumber int `json:"build-number"`
+
+	// Parameters contains all parameters passed
+	// to this build
+	Parameters map[string]string `json:"parameters"`
+}
+
+// GcsPath determines where artifacts for this job should
+// be uploaded to under the base bucket
+func (m *Metadata) GcsPath() string {
+	return m.JobName + "/" + strconv.Itoa(m.BuildNumber)
+}
+
+// uploadFilesToGCS uploads all of the files in the
+// uploadTargets map to GCS in parallel. The map is
+// from filesystem path to GCS path under the bucket
+func uploadFilesToGCS(bucket *storage.BucketHandle, uploadTargets map[string]string) error {
+	errCh := make(chan error, len(uploadTargets))
+	group := &sync.WaitGroup{}
+	group.Add(len(uploadTargets))
+	for src, dest := range uploadTargets {
+		obj := bucket.Object(dest)
+		go func(obj *storage.ObjectHandle, file string) {
+			defer group.Done()
+			if err := uploadFileToGCS(obj, file); err != nil {
+				errCh <- err
+			}
+		}(obj, src)
+	}
+	group.Wait()
+	close(errCh)
+	if len(errCh) != 0 {
+		var uploadErrors []error
+		for err := range errCh {
+			uploadErrors = append(uploadErrors, err)
+		}
+		return fmt.Errorf("encountered errors during upload: %v", uploadErrors)
+	}
+
+	return nil
+}
+
+// uploadFileToGCS copies all data from the file on
+// disk to the GCS object
+func uploadFileToGCS(obj *storage.ObjectHandle, file string) error {
+	reader, err := os.Open(file)
+	if err != nil {
+		return err
+	}
+
+	writer := obj.NewWriter(context.Background())
+	if _, err := io.Copy(writer, reader); err != nil {
+		return err
+	}
+
+	if err := writer.Close(); err != nil {
```

Comment:
@kargakis do you mean closing the writer and the reader? not really sure what you mean